### PR TITLE
feat: add ScoreAnalysis.summarize(int topLimit) method to customize constraint/matches display limit

### PR DESCRIPTION
**Pull request by PatrickDiallo23**
## Summary

This PR adds `ScoreAnalysis.summarize(int topLimit)` method as an alternative to `ScoreAnalysis.summarize()` alternative that shows all constraints/matches  (#1429). 

The method provides a concise summary of constraint matches, limiting the number of matches displayed per constraint to the specified topLimit (that must be > 0). If the number of matches exceeds the limit, the summary includes an ellipsis and a count of the remaining matches.

## Example Behavior

For `topLimit = 2`, the summary will display up to 2 matches per constraint and indicate the remaining count if more matches exist.
For` topLimit = Integer.MAX_VALUE`, all matches are displayed without truncation.

## Tests Added

- **Valid Limits**: Verifies correct behavior for various topLimit values, including edge cases like Integer.MAX_VALUE.
- **Invalid Limits**: Ensures exceptions are thrown for non-positive topLimit values.
- **Ellipsis Verification**: Confirms the summary correctly appends the ellipsis and remaining match count when the limit is exceeded.

### CODE REVIEW

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java [hunk 1]
```diff
@@ -300,6 +300,64 @@ Explanation of score (%s):
...
+     * Provides a summary of the solution's score analysis using the {@link #summarize(int)} method
+     * in a similar way to {@link #summarize()}.
+     * It is possible to specify the maximum number of constraint matches to display per constraint by passing
+     * the desired limit as an argument.
+     *
+     * @param topLimit maximum number of constraint matches to show per constraint.
+     *        Use {@link Integer#MAX_VALUE} to show all matches.
+     */
+    @SuppressWarnings("java:S3457")
+    public @NonNull String summarize(int topLimit) {
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java:275]: Let's not duplicate all of this code.
Surely the existing code can be extracted and generalized so that both the old method and this new method can reuse it.

The old method applies a pre-configured limit. It should not be too hard to implement that method with this method.

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java [hunk 2]
```diff
@@ -234,6 +234,55 @@ Explanation of score (%s):
...
+     * Return a diagnostic text that explains part of the score quality using the {@link #summarize(int)} method
+     * in a similar way to {@link #summarize()}.
+     * It is possible to specify the maximum number of constraint matches to display per constraint by passing
+     * the desired limit as an argument.
+     *
+     * @param topLimit maximum number of constraint matches to display per constraint
+     *        Use {@link Integer#MAX_VALUE} to show all matches.
+     */
+
+    public @NonNull String summarize(int topLimit) {
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java:216]: Dtto.

**Review by triceo — CHANGES_REQUESTED** [review]
Thanks for the contribution!
Will likely be good to go once the code duplication is eliminated.

**Comment by PatrickDiallo23** [comment]
@triceo I refactored the code. Please take a look.

**Comment by PatrickDiallo23** [comment]
I could also just let only the implementation of the summarize(int topLimit) since it can also cover the default one. What do you think?

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java [hunk 3]
```diff
@@ -200,6 +200,28 @@ public String constraintPackage() {
...
+        if (topLimit < 1) {
+            throw new IllegalArgumentException("The topLimit (" + topLimit + ") must be at least 1.");
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java]: We now use a new convention for exception messages with arguments.

    throw new IllegalArgumentException("The topLimit (%d) must be at least 1.".formatted(topLimit));

Provides better readability for long exception messages.
(Not so much for these short ones, but conventions are conventions.)

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java [hunk 4]
```diff
@@ -259,6 +259,28 @@ public Collection<ConstraintAnalysis<Score_>> constraintAnalyses() {
...
+        if (topLimit < 1) {
+            throw new IllegalArgumentException("topLimit (" + topLimit + ") must be positive.");
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java]: Same comment on exception message formatting.

**Review by triceo — APPROVED** [review]
LGTM when comments resolved!

**Comment by PatrickDiallo23** [comment]
@triceo done

**Comment by triceo** [comment]
Thanks for the contribution, @PatrickDiallo23!
